core/src/main/scala/org/apache/spark/scheduler/ShuffleMapStage.scala

```diff
@@ -87,10 +87,23 @@ private[spark] class ShuffleMapStage(
    */
   def isAvailable: Boolean = numAvailableOutputs == numPartitions
 
-  /** Returns the sequence of partition ids that are missing (i.e. needs to be computed). */
+  /**
+   * Returns the sequence of partition ids that are missing (i.e. needs to be computed).
+   * If the current stage is indeterminate, missing partition is all partitions every time.
+   */
   override def findMissingPartitions(): Seq[Int] = {
     mapOutputTrackerMaster
       .findMissingPartitions(shuffleDep.shuffleId)
       .getOrElse(0 until numPartitions)
   }
+
+  /**
+   * Clear the intermediate status including shuffle status and indeterminate attempt id for
+   * shuffle map stage.
+   */
+  override def clearIntermediateState(): Unit = {
+    if (isIndeterminate) {
+      mapOutputTrackerMaster.unregisterAllMapOutput(shuffleDep.shuffleId)
```

Comment:
seems it's not a right abstraction, as only shuffle map stage does cleanup. Can we move this logic to the DAGScheduler?

Reply:
No problem.